src/storage/lsm.rs

```diff
@@ -2,46 +2,52 @@ use super::api::{Key, Value};
 use super::serde;
 use anyhow::Result;
 use std::collections::BTreeMap;
-use std::fs::{File, OpenOptions};
+use std::fs::{self, File, OpenOptions};
 use std::io::{Seek, SeekFrom};
+use std::mem;
 use std::path::PathBuf;
 use std::time::{SystemTime, UNIX_EPOCH};
 
-static COMMIT_LOG_PATH: &'static str = "/tmp/pancake/commit_log.data";
-static SSTABLE_DIR_PATH: &'static str = "/tmp/pancake/sstables";
-static SSTABLE_IDX_SPARSENESS: usize = 4;
-static MEMTABLE_FLUSH_SIZE_THRESH: usize = 3;
+static COMMIT_LOGS_DIR_PATH: &'static str = "/tmp/pancake/commit_logs";
+static SSTABLES_DIR_PATH: &'static str = "/tmp/pancake/sstables";
+static SSTABLE_IDX_SPARSENESS: usize = 3;
+static MEMTABLE_FLUSH_SIZE_THRESH: usize = 7;
 static SSTABLE_COMPACT_COUNT_THRESH: usize = 4;
 
+fn new_path(parent_path: &str) -> PathBuf {
+    let mut path = PathBuf::from(parent_path);
+    path.push(format!(
+        "{}.data",
+        SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_micros()
+    ));
+    path
+}
+
 /// The memtable: in-memory sorted map of the most recently put items.
 /// Its content corresponds to the append-only commit log.
 /// The memtable and commit log will be flushed to a (on-disk SSTable, in-memory sparse seeks of this SSTable) pair, at a later time.
 #[derive(Default, Debug)]
 struct Memtable(BTreeMap<Key, Value>);
 
 impl Memtable {
-    fn read_from_commit_log(path: &PathBuf) -> Result<Self> {
-        let mut memtable = Self::default();
-
-        if !path.exists() {
-            return Ok(memtable);
-        }
-
-        let mut file = File::open(path)?;
+    fn update_from_commit_log(&mut self, path: &PathBuf) -> Result<()> {
```

Comment:
This is used in two places:

1. during initialization
2. sstable compaction

During init, we want to read commit log(s) into the memtable. Previously we had only one commit log file. Now we can have two (commit log, memtable) pairs, or more in case of io error. This came from this PR's modification of compaction so that it could be run concurrently. We 1) swap the main memtable with a "memtable being flushed" (`memtable_in_flush`); 2) compact 3) delete `memtable_in_flush` and delete the commit log. As a result, to read any straggler commit files that might be left behind from a previous shutdown error, we're now reading multiple commit log files at init. (Right now I'm just leaving behind the latest commit log, so there is always one commit log file left at shutdown, in the happy path. I don't think this is the right way for a non-toy db, but meh it's ok for now.)

SSTable compaction: we need 1) to read all sstables (that are being compacted) densely into memory; 2) write the dense in-memory dict to disk; 3) create a sparse version of the in-memory dict. For the dense in-memory dict, I'm reusing the `Memtable` struct.

So in both cases, reading the whole file is necessary (unless I'm missing something). Time is indeed O(sum of all file sizes). But slow init is fine, and once we get async compaction working, it shouldn't affect frontend queries.

Reply:
> So in both cases, reading the whole file is necessary (unless I'm missing something). Time is indeed O(sum of all file sizes). But slow init is fine, and once we get async compaction working, it shouldn't affect frontend queries.

It's not the linear scan of the files I was thinking about. It's the BTree insertions. 

Merging two sorted lists can be done in linear time, but the BTree insertion is `logn`. So instead of a O(n) overall, it is O(n*logn) where `n` is the number of keys/entries.

Reply:
Ah sorry you're right! I guess it will be `O(s)` where `s` is the number of files.

Reply:
So the in-memory index of an SSTable can be a vector right?

Reply:
> I guess it will be O(s) where s is the number of files.

This will become true if a data structure is used which has O(1) insertions or if a merge algorithm is used which can walk along the input files together and incrementally combine the sorted lists.

Reply:
Oh oops I meant `O(log(s))`. We need a priority queue to track where each file is right?

Can I assign it to you? :)

Reply:
It wouldn't be `log(s)` because that would mean that if there are 100 files and log_10 is used as the `log` then only 10 files of 100 need to be inspected. Since all files need to be scanned, it's at least linear in the number of files. 

But we can't claim a bound of O(s) either if a BTree is used for insertion because each insertion is logarithmic with respect to the number of entries in the tree.

So overall, the file level granularity isn't sufficient to express the runtime because each element in the file has a runtime complexity greater than constant.

Probably more explanatory and precise to think in terms of keys.

> Can I assign it to you? :)
 
yeah!

Reply:
Omg I'm gonna fail interviews. I meant to say `O(s log(s))`, but that's not entirely accurate either. I think it's like `O(n log(s))`? Anyway thanks for taking it!

Reply:
I have an idea for the compaction algorithm. I'll put something together and see if it actually works.